### PR TITLE
Allow for SSH URLs when using hg repository type

### PR DESCRIPTION
**Pull request by gaxweb**
This fixes ticket #11874

It was declared a feature, so I've chosen the main branch. I also applied a few minor fixups to docblocks for example. I hope that's ok, otherwise let me know and I'll remove them.

For the fix itself, I basically just adjusted the regex to allow "ssh" and to grab the password more reliably, even though the latter isn't even used. Worked fine with the repo I'm using, if the SSH key is set up in both the repo and the workstation you're issuing the pull from.

### CODE REVIEW

File: src/Composer/Util/Hg.php [hunk 1]
```diff
@@ -93,6 +102,7 @@ private function throwException($message, string $url): void
     /**
      * Retrieves the current hg version.
      *
+     * @param ProcessExecutor $process
```
Comment (stof) [inline src/Composer/Util/Hg.php:105]: this is useless as it brings no additional information compared to the native signature.

File: src/Composer/Util/Hg.php [hunk 2]
```diff
@@ -70,21 +77,23 @@ public function runCommand(callable $commandCallable, string $url, ?string $cwd)
...
-        $this->throwException('Failed to clone ' . $url . ', ' . "\n\n" . $error, $url);
+        $this->throwException("Failed to clone $url, \n\n" . $error, $url);
     }
 
     /**
      * @param non-empty-string $message
-     *
-     * @return never
+     * @param string $url
+     * @return never-returns
```
Comment (stof) [inline src/Composer/Util/Hg.php:89]: this should be reverted. `never` is the native name for this type. there is no reason to use a phpstan-only alias of this type name.

File: src/Composer/Util/Hg.php [hunk 3]
```diff
@@ -70,29 +77,29 @@ public function runCommand(callable $commandCallable, string $url, ?string $cwd)
...
+                'Failed to clone ' . $url . ', hg was not found, check that it is installed and in your PATH env.' . "\n\n" . $this->process->getErrorOutput()
+            ));
         }
 
         throw new \RuntimeException(Url::sanitize($message));
     }
 
     /**
      * Retrieves the current hg version.
-     *
```
[the remark this answers is not in the document]
Comment (fredden) [reply src/Composer/Util/Hg.php:95]: The official documentation says there should be a blank line after the summary and before the tags. https://github.com/php-fig/fig-standards/blob/master/proposed/phpdoc.md#5-the-phpdoc-format

**Comment by Seldaek** [comment]
Ok thanks for the PR. I did some more changes to the regex, hopefully you can check with latest composer snapshot that this is still working well for you.

**Pull request merged**